core/identity-management/iin-agent/src/protocols/externalOperations.ts

```diff
@@ -5,26 +5,224 @@
+    console.log('sendIdentityConfiguration:', remoteSecurityDomain, '-', remoteMemberId, '-', nonce);
+    try {
+        if (attestedMembership.hasError()) {
+            throw new Error(attestedMembership.getError());
+        }
+        if (attestation.getCertificate().length == 0) {
+            throw new Error('attestation has no certificate');
+        }
+        if (attestation.getSignature().length == 0) {
+            throw new Error('attestation has no signature');
+        }
+        securityDomainMap.set(secDomMapKey, attestedMembership);
+    } catch (e) {
+        const errorMsg = `${e} from SecurityDomain: ${remoteSecurityDomain}, Member: ${remoteMemberId}, Nonce: ${nonce}`;
+        console.error(errorMsg);
+        securityDomainMap.set(secDomMapKey, errorMsg);
+    }
+    const currForeignAgentResponsesCount = foreignAgentResponseCount.get(nonce).current;
+    const totalForeignAgentResponsesCount = foreignAgentResponseCount.get(nonce).total;
+    foreignAgentResponseCount.set(nonce, { current: currForeignAgentResponsesCount + 1, total: totalForeignAgentResponsesCount});
+    
+    if (currForeignAgentResponsesCount + 1 < totalForeignAgentResponsesCount) {
+        // Pending respones from other foreign iin-agents;
+        return;
+    }
+    if (currForeignAgentResponsesCount + 1 > totalForeignAgentResponsesCount) {
+        console.warn('Warning: Received extra response.');
+    }
+    
+    // All attesations received, Group Attestations;
+    let attestations = [];
+    let membership = "";
+    let errorMsg = "";
+    const remoteSecurityDomainDNS = utils.getSecurityDomainDNS(remoteSecurityDomain);
+    for (const remoteAgent in remoteSecurityDomainDNS) {
+        const key = getSecurityDomainMapKey(remoteSecurityDomain, remoteAgent, nonce);
+        if (!securityDomainMap.has(key)) {
+            // count completed but still some foreign agents haven't responded.
+            console.log(`Waiting for response from ${remoteAgent}`);
+            return;
+        }
+        let attestedMembershipOrError = securityDomainMap.get(key);
+        if (typeof attestedMembershipOrError === "string") {
+            errorMsg = attestedMembershipOrError as string;
+        }
+        attestedMembershipOrError = attestedMembershipOrError as iin_agent_pb.AttestedMembership;
+        const attestation = attestedMembershipOrError.getAttestation()!;
+        if (membership === "") {
+            membership = attestedMembershipOrError.getMembership();
+        } else if (membership !== attestedMembershipOrError.getMembership()) {
+            errorMsg = `received different membership from ${remoteAgent}`;
         }
+        if (!utils.validateAttestedMembership(membership, nonce, attestation)) {
+            errorMsg = `attested membership from ${remoteAgent} invalid.`;
+        }
+        attestations.push(attestation);
+        securityDomainMap.set(getSecurityDomainMapKey(remoteSecurityDomain, remoteAgent), attestedMembershipOrError);
```

Comment:
Why do we need to manage these memory maps at the sending side (I know why we need them in the receiver)? From what I can see, they are never looked up; only set or deleted.

Reply:
This is for caching the attested membership, so that if some local agent requests counter attestation, this agent can look in cache if it had recently fetched.

That code Shikhar will be writing, its not in this PR.